### PR TITLE
is_obstructed? Method

### PR DESCRIPTION
**Pull request by xJIsaac**
Created is_obstruced?, horizontal_vertical_array, diagonal_array and
contains_piece? methods along with tests. Moved Seb’s logic to game
model as what_at method.

### CODE REVIEW

**Comment by justinrjoseph** [comment]
@j7isaac, I request two things at this time:

1) More descriptive variables: in this PR's case I'm pointing at only one, your 'array' variable in each obstruction method...but in general it's a good practice to declare your variables as meaningful terms. We may know what an array is, but within the context of your methods, I don't have a clue. What _is_ the array, essentially? What is it for?

2) **Comments**!!

Oohh, the dreaded word :P

I understand that many developers groan at the mere thought of writing comments, but those developers make others quite unhappy when they say NOTHING about what their code is doing. People aren't mind-readers, and developers that didn't know your thought processes will waste a lot of time and brain power trying to figure out what you did, and your rationale for it. While specific lines of code might be self-explanatory, the logic behind it isn't necessarily.

In other words, please add some descriptive comments to your methods. Especially for logic as complicated as determining obstruction for chess pieces, it will help the rest of us understand just how your code is legitimately determining whether or not an obstruction exists for a given moving piece.

And don't worry, I'm preaching at myself on this too.

**Comment by xJIsaac** [comment]
@justinrjoseph Added comments.

**Comment by xJIsaac** [comment]
@justinrjoseph Btw, while I appreciated the critique and agreed with most of it, I'm not entirely sure if I agree with this statement, "those developers make others quite unhappy when they say NOTHING about what their code is doing."

Here, are some good reads which argue the use of comments as a last resort.
http://blog.codinghorror.com/coding-without-comments/
http://www.codeodor.com/index.cfm/2008/6/18/Common-Excuses-Used-To-Comment-Code-and-What-To-Do-About-Them/2293

In the end it can be considered a stylistic preference and not something every developer would be unhappy about if missing.

Just trying to be objective and helpful. :P

**Comment by justinrjoseph** [comment]
@j7isaac I'm not interested in joining what I believe is an unprofitable debate. I'm sure I could just now Google articles that justify just as well the usefulness of comments. I fall in the realm that some commenting is necessary, but not all...which seems to be what your resources conclude anyway. My point ultimately was just to provide _something_ for our obstruction code because I don't believe it's something you can easily wrap your head around in a few minutes. I wasn't requesting a novel. I even think some of the comments you've provided aren't necessary, but I have no reason to complain that they're there. So I thank you for your willingness in this way, and I think it will benefit the group in the long haul.
